### PR TITLE
Closure elimination for new backend

### PR DESCRIPTION
**Pull request by lrytz**
When an indylambda closure is allocated and invoked within the same method, rewrite the invocation to the implementation method.

### CODE REVIEW

**Comment by lrytz** [comment]
Ready for a first round of feedback. @retronym, @SethTisue

**Comment by retronym** [comment]
I gave this an in-person review on Tuesday. I thought @lrytz added most of my comments to as GitHub comments, but I don't see them here. AFAICT, the first commit has incorporated my comments, but the second one still needs to be polished.

The tests for prod/cons analysis are nicely structured.

The IDE build failure is unrelated to this patch:

```
[sbt-republish] [info] Updating {file:/home/jenkins/workspace/scala-2.11.x-integrate-ide/target/zinc/target-0.9.1/project-builds/sbt-republish-40094874e9a4cc023749c371f693ef67a5da2d41/project/project/}sbt-republish-40094874e9a4cc023749c371f693ef67a5da2d41-build-build...
[sbt-republish] [warn]  [FAILED     ] org.apache.maven#maven-compat;3.0.4!maven-compat.jar: invalid sha1: expected=14d626e1e29a36f5f4629b689e964f8665707839 computed=211dc5b40b735d75fd7ac245a0ca5e82a3ab1508 (3637ms)
[sbt-republish] [warn]  [FAILED     ] org.apache.maven#maven-compat;3.0.4!maven-compat.jar: invalid sha1: expected=14d626e1e29a36f5f4629b689e964f8665707839 computed=211dc5b40b735d75fd7ac245a0ca5e82a3ab1508 (3637ms)
[sbt-republish] [warn] ==== public: tried
[sbt-republish] [warn]   http://repo1.maven.org/maven2/org/apache/maven/maven-compat/3.0.4/maven-compat-3.0.4.jar
[sbt-republish] [warn]  [FAILED     ] org.jsoup#jsoup;1.6.1!jsoup.jar: invalid sha1: expected=70f0dfbe511d84ee7d0dda238a52746751787c47 computed=f124031a316ad95c6498ca5d97e7bcf0fffa00d1 (1941ms)
[sbt-republish] [warn]  [FAILED     ] org.jsoup#jsoup;1.6.1!jsoup.jar: invalid sha1: expected=70f0dfbe511d84ee7d0dda238a52746751787c47 computed=f124031a316ad95c6498ca5d97e7bcf0fffa00d1 (1941ms)
```

**Comment by lrytz** [comment]
@retronym indeed, i cleaned up the first commit, but still working on the second. should appear soon. my self-comments are in this revision: https://github.com/lrytz/scala/commit/0b7e0f108bc2406933c45e2fd70912543410c34b

**Comment by lrytz** [comment]
@retronym done with the cleanup. Summary of changes since last review:

ProdCons
- ProdCons correctly handles instructions that produce multiple values (e.g., `DUP`).
- ProdCons supports all exotic copying instructions (e.g., `DUP2_X2`) for queries that thread through copying instructions (`initialProducers..` / `ultimateConsumers..`)
- fixed ProdCons for `INSTANCEOF`
- use `AnyRefMap` for mutable maps
- more testing

Closure apply rewriting
- created `Handle` objects to compare against for `metafactory` / `altMetafactory` (instead of comparing all fields). Similar for the implMethod, created a Type to compare against. See `isClosureInstantiation`.
- cleaned up the trickery with the opCode-offset, better comments
- run ProdCons only if it's actually needed, i.e., if there's some callsite within the method that looks like a closure apply call (in terms of name and descriptor).
- support all calling conventions (`NEWINVOKESPECIAL` etc)

File: src/compiler/scala/tools/nsc/backend/jvm/BackendReporting.scala [hunk 1]
```diff
@@ -247,6 +247,28 @@ object BackendReporting {
                                      callsiteClass: InternalName, callsiteName: String, callsiteDesc: String) extends CannotInlineWarning
 
   /**
+   * Used in `rewriteClosureApplyInvocations` when a closure apply callsite cannot be rewritten
+   * to the closure body method.
+   */
+  trait RewriteClosureApplyToClosureBodyFailed extends OptimizerWarning {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/BackendReporting.scala:253]: sealed?

**Comment by retronym** [comment]
Regarding the using or pre-allocated tuples in `InstructionStackEffect` (`def t`), I think that `IntMap#contains` actually allocates an `Option` as it is implemented in terms of `get(k).isDefined`.

**Comment by retronym** [comment]
LGTM as a starting point. We'll need to add automated tests once we have the indylambda functional interfaces in the repo, which should do this week, using your patch to embed the `JFunctionN` sources.

Have you performed any manual testing with indylambda and closure inlining enabled for a big codebase (e.g. boostrapping scala)?

**Comment by lrytz** [comment]
good piont about `IntMap`.. Running bootstrap testing now.

**Comment by lrytz** [comment]
found one issue:

```
    val fun: String => String = l => l
    val l = List("")
    fun(l.head)
```

Here, `l.head` has type `Object`, but the lambda body function expects a `String`, so I get

```
java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    Test$.f()Ljava/lang/String; @42: invokestatic
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to 'java/lang/String'
```

**Comment by retronym** [comment]
I guess we'll need the type adaptation framework after all.

**Comment by lrytz** [comment]
i think we can just cast in these cases

**Comment by lrytz** [comment]
Added another commit. The test suite seems to run fine now (some unrelated / expected failures due to indyLambda, closure rewriting). Also `strap.done` / `test.stability` work.

**Comment by retronym** [comment]
LGTM! Great work!

**Pull request merged**